src/include/OSL/mask.h

```diff
@@ -0,0 +1,370 @@
+// But be careful if you ever have two Masked::LaneProxy's with
+// different masks
+class ActiveLane : public Lane {
+public:
+    explicit OSL_FORCEINLINE ActiveLane(int index) : Lane(index) {}
+
+    ActiveLane() = delete;
+
+    OSL_FORCEINLINE ActiveLane(const ActiveLane& other) : Lane(other) {}
+};
+
+template<int WidthT> class Mask {
+    typedef unsigned short Value16Type;
+    static_assert(sizeof(Value16Type) == 2, "unexpected platform");
+
+    typedef unsigned int Value32Type;
+    static_assert(sizeof(Value32Type) == 4, "unexpected platform");
+
+    typedef unsigned long long Value64Type;
+    static_assert(sizeof(Value64Type) == 8, "unexpected platform");
+
+    typedef
+        typename std::conditional<WidthT <= 32, Value32Type, Value64Type>::type
+            Value32or64Type;
+
+public:
+#if 0  // Enable 16bit integer storage of masks, vs 32bit.
+    typedef typename std::conditional<WidthT <= 16,
+                                      Value16Type,
+                                      Value32or64Type>::type ValueType;
+#else
+    typedef Value32or64Type ValueType;
+#endif
+
+    static constexpr int width = WidthT;
+
+protected:
+    static constexpr int value_width = sizeof(ValueType) * 8;
+    static_assert(value_width >= WidthT, "unsupported WidthT");
+    static constexpr ValueType valid_bits
+        = static_cast<ValueType>(0xFFFFFFFFFFFFFFFF) >> (value_width - WidthT);
+
+public:
+    OSL_FORCEINLINE Mask() {}
+
+    explicit OSL_FORCEINLINE Mask(Lane lane) : m_value(1 << lane.value()) {}
+
+    explicit OSL_FORCEINLINE Mask(bool all_on_or_off)
+        : m_value((all_on_or_off) ? valid_bits : 0)
+    {
+    }
+
+    explicit constexpr OSL_FORCEINLINE Mask(std::false_type) : m_value(0) {}
+
+    explicit constexpr OSL_FORCEINLINE Mask(std::true_type)
+        : m_value(valid_bits)
+    {
+    }
+
+    explicit OSL_FORCEINLINE Mask(Value16Type value_)
+        : m_value(static_cast<ValueType>(value_))
+    {
+    }
+
+    explicit OSL_FORCEINLINE Mask(Value32Type value_)
+        : m_value(static_cast<ValueType>(value_))
+    {
+    }
+
+    explicit OSL_FORCEINLINE Mask(Value64Type value_)
+        : m_value(static_cast<ValueType>(value_))
+    {
+    }
+
+    explicit OSL_FORCEINLINE Mask(int value_)
+        : m_value(static_cast<ValueType>(value_))
+    {
+    }
+
+    OSL_FORCEINLINE Mask(const Mask& other) : m_value(other.m_value) {}
+
+    template<int OtherWidthT,
+             typename = pvt::enable_if_type<(OtherWidthT < WidthT)>>
+    explicit OSL_FORCEINLINE Mask(const Mask<OtherWidthT>& other)
+        : m_value(static_cast<ValueType>(other.value()))
+    {
+    }
+
+
+    OSL_FORCEINLINE ValueType value() const { return m_value; }
+
+    // count number of active bits
+    OSL_FORCEINLINE int count() const
+    {
+#if __INTEL_COMPILER
+        if (value_width <= 32)
+            return _mm_popcnt_u32(m_value);
+        else
+            return _mm_popcnt_u64(m_value);
+#else
```

Comment:
Should we have a clause for gcc and clang to use the __builtin_popcount ?

Reply:
Feel free to add that, I didn't have all the gcc test matrix setup locally, does clang have a similar builtin?  However if gcc and clang recognize the idiom, that might be more maintainable.  Will have to test in godbolt.org

Reply:
Verified idiom is not detected, correct code though.  I'll try adding case for clang, gcc, then else to the idiom

Reply:
Looks like `__builtin_popcount` works with both gcc and clang, and ought to be able to be counted on to do the best thing on any architecture that those compilers support.

Reply:
I stand corrected, gcc and clang both recognized the idiom when targeting an ISA that supports popcount instruction.  I'll update it to be compiler intrinsics anyway, with a comment to use std::bitcount for c++20 in the future.

Reply:
Please include a comment

    // FIXME(C++20)

as a breadcrumb that will let us find those spots when it comes time to fix things up nicely for C++20.